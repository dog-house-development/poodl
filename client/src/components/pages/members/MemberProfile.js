import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import _ from 'lodash';
import MemberActions from '../../../actions/memberActions';
import Loading from '../../ui/Loading';
import EditableField from '../../ui/EditableField';
import EditableRadio from '../../ui/EditableRadio';
import EditableCheckBox from '../../ui/EditableCheckBox';
import { Link } from 'react-router-dom';
import Button from '../../ui/Button';
import memberFields, { Categories } from './memberFields';

export class MemberProfile extends Component {
    constructor(props) {
        super(props);
        this.routeParam = props.match.params.id;
        this.state = {
            editMode: _.fromPairs(_.map(_.values(Categories), category => [category.id, false])),
            fields: _.fromPairs(
                _.map(memberFields, field => (field.type === 'checkbox' ? [field.id, false] : [field.id, '']))
            ),
            modifiedFields: {}
        };
        this.handleFieldChange = this.handleFieldChange.bind(this);
        this.handleEditClick = this.handleEditClick.bind(this);
    }

    componentDidMount() {
        // call redux action to retrieve specified member from api
        this.props.memberActions.get(this.routeParam);
    }

    componentDidUpdate(prevProps, prevState) {
        if (!_.isEqual(prevProps.member, this.props.member)) {
            this.setState({
                fields: {
                    ...this.props.member
                }
            });
        }
    }

    handleFieldChange(e) {
        if (e.target.type === 'checkbox') {
            this.setState({
                fields: { ...this.state.fields, [e.target.id]: e.target.checked },
                modifiedFields: { ...this.state.modifiedFields, [e.target.id]: e.target.checked }
            });
        } else {
            this.setState({
                fields: { ...this.state.fields, [e.target.id]: e.target.value },
                modifiedFields: { ...this.state.modifiedFields, [e.target.id]: e.target.value }
            });
        }
    }

    handleEditClick(e) {
        e.preventDefault();
        if (this.state.editMode[e.target.id] && !_.isEmpty(this.state.modifiedFields)) {
            this.props.memberActions.edit(this.routeParam, this.state.modifiedFields);
            this.setState({ modifiedFields: {} });
        }
        this.setState({ editMode: { [e.target.id]: !this.state.editMode[e.target.id] } });
    }

    getFieldsMarkup(value, key) {
        let fieldsMarkup = [];

        fieldsMarkup.push(
            <div key={key} className="profile-section-header">
                <div className="profile-section-title-and-description">
                    <h3 className="profile-section-title">{value.title}</h3>
                    {value.description ? <p>{value.description}</p> : ''}
                </div>
                <div className="profile-section-button">
                    <Button
                        key={_.uniqueId('edit-button-')}
                        content={this.state.editMode[key] ? 'Done' : 'Edit'}
                        id={Categories[key].id}
                        onClick={this.handleEditClick}
                        size="small"
                    />
                </div>
            </div>
        );
        fieldsMarkup.push(
            _.map(_.filter(memberFields, field => field.category === key), field => {
                if (field.type === 'checkbox') {
                    return (
                        <EditableCheckBox
                            id={field.id}
                            key={field.id}
                            defaultValue={this.state.fields[field.id]}
                            editMode={this.state.editMode[field.category]}
                            handleChange={this.handleFieldChange}
                            label={field.label}
                        />
                    );
                } else if (field.type === 'radio') {
                    return (
                        <EditableRadio
                            id={field.id}
                            key={field.id}
                            defaultValue={this.state.fields[field.id]}
                            options={field.options}
                            editMode={this.state.editMode[field.category]}
                            handleChange={this.handleFieldChange}
                            label={field.label}
                        />
                    );
                } else {
                    return (
                        <EditableField
                            id={field.id}
                            key={field.id}
                            defaultValue={this.state.fields[field.id]}
                            editMode={this.state.editMode[field.category]}
                            handleChange={this.handleFieldChange}
                            label={field.label}
                        />
                    );
                }
            })
        );

        return fieldsMarkup;
    }

    getPanelMarkup() {
        let panelMarkup = [];
        _.forEach(Categories, (value, key) => {
            panelMarkup.push(
                <div className="panel" key={key}>
                    {this.getFieldsMarkup(value, key)}
                </div>
            );
        });
        return panelMarkup;
    }

    render() {
        return (
            <div className="view-all-container">
                <Link to="/members" className="button small tertiary">
                    <i className="material-icons">keyboard_backspace</i> Back to all members
                </Link>
                <div>
                    <h1>
                        {this.props.loading ? <Loading /> : _.get(this.props.member, 'firstName')}{' '}
                        {this.props.loading ? '' : _.get(this.props.member, 'lastName')}
                    </h1>
                    <div>{this.getPanelMarkup()}</div>
                </div>
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        member: state.members.all[props.match.params.id],
        loading: state.members.loading,
        errors: state.members.errors
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        memberActions: bindActionCreators(MemberActions, dispatch)
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(MemberProfile);
