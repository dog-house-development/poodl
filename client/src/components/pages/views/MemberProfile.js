import React, { Component } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import { fetchMember } from '../../../actions/memberActions';
import Loading from '../../ui/Loading';
import { Link } from 'react-router-dom';

export class MemberProfile extends Component {
    constructor(props) {
        super(props);
        this.routeParam = props.match.params.id;
    }

    componentDidMount() {
        // call redux action to retrieve specified member from api
        this.props.getMember(this.routeParam);
    }

    getPageMarkup() {
        if (this.props.loading) {
            return <Loading content="Loading admin info..." />;
        } else {
            return (
                <div>
                    <h1>
                        {_.get(this.props.member, 'firstName')} {_.get(this.props.member, 'lastName')}
                    </h1>
                    <div className="panel">
                        {_.map(_.toPairs(this.props.member), pair => {
                            if (pair[0] === 'password' || pair[0] === '__v') {
                                return;
                            }
                            if (pair[0] === '') {
                                pair[1] = _.get(this.props.member, 'superAdmin') ? 'Yes' : 'No';
                            }
                            if (pair[0] === 'date' || pair[0] === 'createdAt' || pair[0] === 'updatedAt') {
                                pair[1] = moment(pair[1]).format('MMMM Do, YYYY, h:mm:ss a');
                            }
                            return (
                                <div key={pair[0]} style={{ margin: '10px 0' }}>
                                    <h4 style={{ color: '#a5a9af', fontSize: '0.8em' }}>{_.upperCase(pair[0])}</h4>
                                    <p>{pair[1]}</p>
                                </div>
                            );
                        })}
                    </div>
                </div>
            );
        }
    }

    render() {
        return (
            <div className="view-all-container">
                <Link to="/members" className="button small tertiary">
                    <i className="material-icons">keyboard_backspace</i> Back to all members
                </Link>
                <div className="panel">{this.getPageMarkup()}</div>
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        member: state.members.one,
        errors: state.errors
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        getMember: id => dispatch(fetchMember(id))
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(MemberProfile);
