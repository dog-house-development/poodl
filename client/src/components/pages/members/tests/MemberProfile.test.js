import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import _ from 'lodash';
import { Link } from 'react-router-dom';
import { MemberProfile, mapStateToProps, mapDispatchToProps } from '../MemberProfile';

configure({ adapter: new Adapter() });

const match = {
    params: {
        _id: '123' //any id you want to set
    }
};
describe('MemberProfile tests', () => {
    let state, props, wrapper, instance;
    const setInstanceAndWrapper = (_props = {}, _state = {}) => {
        state = _.assign(
            {},
            {
                members: {
                    all: {
                        123: {
                            _id: '123',
                            firstName: 'Big',
                            lastName: 'Tup',
                            membershipDate: '02/20/2020',
                            email: 'bigtup@nowhere.com'
                        }
                    },
                    errors: {},
                    loading: false
                },
                services: {
                    all: {
                        123: { memberId: '123' },
                        321: { memberId: '444' }
                    }
                }
            },
            _state
        );
        props = _.assign({}, { match: { params: { id: '123' } } }, _props);
        wrapper = shallow(
            <MemberProfile
                match={match}
                {..._.assign(
                    {},
                    props,
                    mapStateToProps(state, props),
                    mapDispatchToProps(jasmine.createSpy('dispatch'))
                )}
            />
        );
        instance = wrapper.instance();
    };

    beforeEach(() => {
        setInstanceAndWrapper();
    });

    describe('mapStateToProps', () => {
        it('should map state to props', () => {
            expect(mapStateToProps(state, props)).toEqual({
                member: {
                    _id: '123',
                    firstName: 'Big',
                    lastName: 'Tup',
                    membershipDate: '02/20/2020',
                    email: 'bigtup@nowhere.com'
                },
                loading: false,
                errors: {},
                serviceCount: 1
            });
        });
    });

    describe('mapDispatchToProps', () => {
        it('should map dispatch to props', () => {
            const dispatch = jest.fn();
            expect(JSON.stringify(mapDispatchToProps(dispatch))).toEqual(
                JSON.stringify({ memberActions: {}, serviceActions: {} })
            );
        });
    });

    describe('render', () => {
        it('should render correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
