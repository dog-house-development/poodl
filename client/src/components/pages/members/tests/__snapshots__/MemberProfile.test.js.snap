// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberProfile tests render should render correctly 1`] = `
<div
  className="view-all-container"
>
  <Link
    className="button small tertiary"
    replace={false}
    to="/members"
  >
    <i
      className="material-icons"
    >
      keyboard_backspace
    </i>
     Back to all members
  </Link>
  <div>
    <h1>
      Big
       
      Tup
    </h1>
    <DynamicForm
      editValues={[Function]}
      editable={true}
      errors={Object {}}
      getValues={[Function]}
      inputs={
        Array [
          Object {
            "id": "Personal Info",
            "inputs": Array [
              Object {
                "id": "name",
                "inputs": Array [
                  Object {
                    "id": "firstName",
                    "kind": "field",
                    "label": "First name",
                    "type": "text",
                  },
                  Object {
                    "id": "lastName",
                    "kind": "field",
                    "label": "Last name",
                    "type": "text",
                  },
                ],
                "kind": "flex",
              },
              Object {
                "id": "birthDate",
                "kind": "datePicker",
                "label": "Date of Birth",
                "maxYear": 1990,
                "minYear": 1900,
              },
              Object {
                "id": "email",
                "kind": "field",
                "label": "Email",
                "type": "email",
              },
              Object {
                "id": "address",
                "kind": "field",
                "label": "Address",
                "type": "text",
              },
              Object {
                "id": "gender",
                "kind": "radio",
                "label": "Gender",
                "options": Array [
                  "male",
                  "female",
                  "other",
                ],
              },
            ],
            "kind": "group",
          },
          Object {
            "id": "Emergency Contact Info",
            "inputs": Array [
              Object {
                "id": "emergencyContactName",
                "kind": "field",
                "label": "Emergency Contact Name",
                "type": "text",
              },
              Object {
                "id": "emergencyContactRelationship",
                "kind": "field",
                "label": "Emergency Contact Relationship",
                "type": "text",
              },
              Object {
                "id": "emergencyContactPhoneNumber",
                "kind": "field",
                "label": "Emergency Contact Phone Number",
                "type": "text",
              },
            ],
            "kind": "group",
          },
          Object {
            "id": "Membership Details",
            "inputs": Array [
              Object {
                "id": "memberisNewOrRenewal",
                "kind": "radio",
                "label": "Member is new or renewal",
                "options": Array [
                  "new",
                  "renewal",
                ],
              },
              Object {
                "id": "formOfPayment",
                "kind": "radio",
                "label": "Method of payment",
                "options": Array [
                  "cash",
                  "check",
                  "card",
                  "other",
                ],
              },
              Object {
                "id": "bankCheckNumber",
                "kind": "field",
                "label": "Bank Check Number",
                "type": "text",
              },
              Object {
                "id": "includedInEstatePlans",
                "kind": "selectBoolean",
                "label": "Included in Estate Plans",
              },
              Object {
                "id": "wantsEstateInfo",
                "kind": "selectBoolean",
                "label": "Member Wants Estate Info",
              },
              Object {
                "id": "membershipDate",
                "kind": "datePicker",
                "label": "Membership Date",
              },
              Object {
                "id": "renewalDate",
                "kind": "datePicker",
                "label": "Renewal Date",
              },
            ],
            "kind": "group",
          },
          Object {
            "id": "Health Info",
            "inputs": Array [
              Object {
                "id": "specialDiet",
                "kind": "field",
                "label": "Special Diet",
                "type": "text",
              },
              Object {
                "id": "medicalIssues",
                "kind": "field",
                "label": "Medical Issues",
                "type": "text",
              },
              Object {
                "id": "disabilities",
                "kind": "field",
                "label": "Disabilities",
                "type": "text",
              },
              Object {
                "id": "mealPreference",
                "kind": "field",
                "label": "Meal Preference",
                "type": "text",
              },
            ],
            "kind": "group",
          },
          Object {
            "id": "Demographic Info",
            "inputs": Array [
              Object {
                "id": "race",
                "kind": "field",
                "label": "Race",
                "type": "text",
              },
              Object {
                "id": "ethnicity",
                "kind": "radio",
                "label": "Ethnicity",
                "options": Array [
                  "Non-Hispanic/Latino",
                  "Hispanic/Latino",
                ],
              },
              Object {
                "id": "numberInHousehold",
                "kind": "field",
                "label": "Number in Household",
                "type": "text",
              },
              Object {
                "id": "isPersonCaregiver",
                "kind": "selectBoolean",
                "label": "If number in household is 2 or more, is one person a caregiver?",
              },
              Object {
                "id": "monthlyIncome",
                "kind": "radio",
                "label": "Monthly Household Income",
                "options": Array [
                  "Under $1,012",
                  "Under $1,372",
                  "Under $1,732",
                  "Under $2,092",
                  "Under $2,452",
                  "Under $2,812",
                  "Over $3,172",
                ],
              },
              Object {
                "id": "disabled",
                "inputs": Array [
                  Object {
                    "id": "isDisabled",
                    "kind": "selectBoolean",
                    "label": "Disabled",
                  },
                  Object {
                    "id": "isVeteran",
                    "kind": "selectBoolean",
                    "label": "Veteran",
                  },
                  Object {
                    "id": "isSpouse60",
                    "kind": "selectBoolean",
                    "label": "If under 60, is spouse over 60?",
                  },
                ],
                "kind": "flex",
              },
              Object {
                "id": "isDisabled60",
                "kind": "selectBoolean",
                "label": "If under 60, are you disabled and living with someone over 60?",
              },
            ],
            "kind": "group",
          },
          Object {
            "description": "If client is receiving services under National Family Caregiver Support Program, complete the following:",
            "id": "Nation Family Caregiver Support Program",
            "inputs": Array [
              Object {
                "id": "caregiver",
                "kind": "radio",
                "label": "Caregiver",
                "options": Array [
                  "Husband",
                  "Wife",
                  "Daughter/DIL",
                  "Son/SIL",
                  "Other Relative",
                  "Other",
                ],
              },
              Object {
                "id": "grandparent",
                "kind": "radio",
                "label": "Grandparent",
                "options": Array [
                  "Grandparent",
                  "Other elder relative",
                  "Other elder non relative",
                ],
              },
              Object {
                "id": "numberOfKidsUnder19",
                "kind": "field",
                "label": "Number of kids younger than 19",
              },
              Object {
                "id": "needsAADL",
                "kind": "multiCheckbox",
                "label": "Needs Assistance with Activities of Daily Living (ADLs)",
                "options": Array [
                  "Eating",
                  "Dressing",
                  "Bathing",
                  "Toileting",
                  "Transferring",
                  "Walking",
                ],
              },
              Object {
                "id": "needsIADL",
                "kind": "multiCheckbox",
                "label": "Needs Assistance with Instrumental Activities of Daily Living (IADLs)",
                "options": Array [
                  "Meal preparation",
                  "Money management",
                  "Shopping",
                  "Transportation",
                  "Telephone use",
                  "Medication management",
                  "Light housework",
                  "Heavy housework",
                ],
              },
            ],
            "kind": "group",
          },
        ]
      }
      loading={false}
      values={
        Object {
          "_id": "123",
          "email": "bigtup@nowhere.com",
          "firstName": "Big",
          "lastName": "Tup",
          "membershipDate": "02/20/2020",
        }
      }
    />
  </div>
</div>
`;
